File: BinarySearch/Python/033SearchInRotatedSortedArray.py
# Time Complexity: O(Logn)
# Space Complexity: O(1)

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l = 0
        r = len(nums) - 1

        while (l <= r):
            m = ((r - l) // 2) + l
            # Start the conditions
            if target == nums[m]:
                return m

            if nums[l] <= nums[m]: # Left half is sorted (we in the left sorted array):
                if target > nums[m] or target < nums[l]:
                    l = m + 1  # Target is in the right half
                else:
                    r = m - 1  # Target is in the left half
            
            else: # Right half is sorted
                if target < nums[m] or target > nums[r]:
                    r = m - 1 # Target is in the left half
                else:
                    l = m + 1 # Target is in the right half
        return - 1
 

File: BinarySearch/C++/033SearchInRotatedSortedArray.cpp
/* 
    Search an element in a rotated sorted array 
    Time complexity : O(Logn)
    Space complexity : O(1)
*/

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0;
        int r = nums.size() - 1;
        
        // Start binary search
        while (l <= r){
            int m = ((r - l) / 2) + l;
            // If midpoint is the target
            if (nums[m] == target){
                return m;
            }
            
            if (nums[l] <= nums[m]){ 
                // we in the left sorted array. The use of the equality check handles 
                // the cases where the left half consists of only one element
                if(target > nums[m] || target < nums[l]){
                   l = m + 1; // Target is in the right sorted array 
                } else {
                    r = m - 1; // Tatget is in the left sorted array
                }
            } else {
                // we in the right sorted array
                if(target < nums[m] || target > nums[r]){
                    r = m - 1; // Target is in the left sorted array
                } else {
                    l = m + 1; // Target is in the right sorted array
                }
            } 
        }

        return -1; 
    }
};